Sources/Sampler/vmc_sampling.cc

```diff
@@ -179,6 +180,307 @@ VectorXcd GradientOfVariance(const Result &result, AbstractMachine &psi,
+      if (++i_ * sampler_.BatchSize() != samples_.rows()) {
+        sampler_.Next();
+      }
+    }
+  };
+
+  detail::LoopV2(steps, Skip{sampler},
+                 Record{sampler, samples, values, gradients, 0});
+  if (gradients.has_value()) detail::SubtractMean(*gradients);
+  return std::make_tuple(std::move(samples), std::move(values),
+                         std::move(gradients));
+}
+
+namespace detail {
+/// A helper class for forward propagation of batches through machines.
+struct Forward {
+  Forward(RbmSpinV2& m, Index batch_size)
+      : machine_{m},
+        X_(batch_size, m.Nvisible()),
+        Y_(batch_size),
+        coeff_(batch_size),
+        i_{0} {}
+
+  /// \brief Returns whether internal buffer if full.
+  bool Full() const noexcept { return i_ == BatchSize(); }
+  /// \brief Returns whether internal buffer if empty.
+  bool Empty() const noexcept { return i_ == 0; }
+  Index BatchSize() const noexcept { return Y_.size(); }
+
+  /// \brief Add an element to internal buffer.
+  ///
+  /// Buffer should not be full!
+  void Push(Eigen::Ref<const Eigen::VectorXd> v, const ConnectorRef& conn) {
+    assert(!Full());
+    X_.row(i_) = v.transpose();
+    for (auto j = Index{0}; j < conn.tochange.size(); ++j) {
+      X_(i_, conn.tochange[j]) = conn.newconf[j];
+    }
+    coeff_(i_) = conn.mel;
+    ++i_;
+  }
+
+  /// \brief Fills the remaining part of internal buffer with visible
+  /// configuration \p v.
+  void Fill(Eigen::Ref<const Eigen::VectorXd> v) {
+    assert(!Empty() && !Full());
+    const auto n = BatchSize() - i_;
+    X_.block(i_, 0, n, X_.cols()) = v.transpose().colwise().replicate(n);
+    coeff_.segment(i_, n).setConstant(0.0);
+    i_ += n;
+    assert(Full());
+  }
+
+  /// \brief Runs forward propagation.
+  ///
+  /// Buffer should be full!
+  std::tuple<const Eigen::VectorXcd&, Eigen::VectorXcd&> Propagate() {
+    assert(Full());
+    machine_.LogVal(X_, /*out=*/Y_, /*cache=*/any{});
+    i_ = 0;
+    return std::tuple<const Eigen::VectorXcd&, Eigen::VectorXcd&>{coeff_, Y_};
+  }
+
+ private:
+  RbmSpinV2& machine_;
+  RowMatrix<double> X_;
+  Eigen::VectorXcd Y_;
+  Eigen::VectorXcd coeff_;
+  Index i_;
+};
+
+struct Accumulator {
+  Accumulator(Eigen::VectorXcd& loc, Forward& fwd)
+      : locals_{loc}, index_{0}, accum_{0.0, 0.0}, forward_{fwd}, states_{} {
+    states_.reserve(forward_.BatchSize());
+  }
+
+  void operator()(Eigen::Ref<const Eigen::VectorXd> v,
+                  const ConnectorRef& conn) {
+    assert(!forward_.Full());
+    forward_.Push(v, conn);
+    ++states_.back().first;
+    if (forward_.Full()) {
+      ProcessBatch();
+    }
+    assert(!forward_.Full());
+  }
+
+  void operator()(Complex log_val) {
+    assert(!forward_.Full());
+    states_.emplace_back(0, log_val);
+  }
+
+  /// Number of visible configurations we've processed is not necessarily a
+  /// multiple of batch size.
+  void Finalize(Eigen::Ref<const Eigen::VectorXd> v) {
```

Comment:
The logic of this can be greatly simplified just imposing that the number of samples must be a multiple of the batch size. Worst-case-scenario, we print a warning saying that the number of samples has been rounded to the nearest multiple of batch size